File: packages/edl/src/tokens.js
/**
 * Token normalisation.
 *
 * Alignment between the original transcript and the user's edited text has to
 * ignore differences that carry no audio consequence: case, punctuation,
 * smart quotes, and the leading space that ASR engines attach to each word.
 * Two tokens that normalise to the same key are considered the same spoken
 * word, so the original audio for it can be kept verbatim.
 */

/** Characters that are punctuation for our purposes but may live inside a word. */
const INNER_PUNCT = /[‘’ʼ']/g; // curly + straight apostrophes
const OUTER_PUNCT = /[^\p{L}\p{N}]+/gu;

/**
 * Reduce a surface word to its comparison key.
 * @param {string} raw
 * @returns {string} normalised key, possibly `''` for punctuation-only input
 */
export function normalizeToken(raw) {
  if (typeof raw !== 'string') return '';
  return raw
    .normalize('NFKC')
    .toLowerCase()
    .replace(INNER_PUNCT, '')
    .replace(OUTER_PUNCT, '')
    .trim();
}

/**
 * Split free text into surface words, preserving the original spelling so that
 * inserted text can be sent to the synthesiser exactly as the user typed it.
 * @param {string} text
 * @returns {string[]}
 */
export function tokenizeText(text) {
  if (!text) return [];
  return text
    .replace(/\s+/g, ' ')
    .trim()
    .split(' ')
    .filter((t) => t.length > 0 && normalizeToken(t).length > 0);
}

File: services/backend/edl/tokens.py
"""Token normalisation.

Alignment between the original transcript and the user's edited text has to
ignore differences that carry no audio consequence: case, punctuation, smart
quotes, and the leading space that ASR engines attach to each word. Two tokens
that normalise to the same key are considered the same spoken word, so the
original audio for it can be kept verbatim.

This must agree exactly with ``voxdocs.synth.normalize_token`` in the model
server, or a word the API believes is in the voice bank will not be found there.
"""

from __future__ import annotations

import unicodedata

#: Curly and straight apostrophes, which live inside words rather than around them.
_APOSTROPHES = "‘’ʼ'"


def normalize_token(raw: str) -> str:
    """Reduce a surface word to its comparison key.

    Returns ``''`` for punctuation-only input, which callers treat as "not a
    word" rather than as an empty word.
    """
    if not raw:
        return ""
    text = unicodedata.normalize("NFKC", str(raw)).lower()
    for mark in _APOSTROPHES:
        text = text.replace(mark, "")
    return "".join(ch for ch in text if ch.isalnum())


def tokenize_text(text: str) -> list[str]:
    """Split free text into surface words.

    The original spelling is preserved so inserted text reaches the synthesiser
    exactly as the user typed it; only the comparison key is normalised.
    """
    if not text:
        return []
    return [word for word in str(text).split() if normalize_token(word)]
